conda_smithy/schema.py

```diff
@@ -0,0 +1,1310 @@
+# This model is also used for generating and automatic documentation for the conda-forge.yml file. The documentation is generated using sphinx and "pydantic-autodoc" extension. For an upstream preview of the documentation, see https://conda-forge.org/docs/maintainer/conda_forge_yml.html.
+
+from enum import Enum, EnumMeta
+from typing import Any, Dict, List, Literal, Optional, Type, Union
+
+from pydantic import BaseModel, Field, create_model
+import yaml
+
+
+class Nullable(Enum):
+    """Created to avoid issue with schema validation of null values in lists or dicts."""
+
+    null = None
+
+
+class PydanticModelGenerator:
+    """
+    A utility class for generating Pydantic models based on an Enum for keys and a Pydantic model for values.
+
+    This class is designed to help mitigate an issue when working with JSON Schema and Pydantic where Enumerators
+    and Dicts do not work well with the generated model. It allows you to dynamically create Pydantic models for
+    cases where you need to map enum values to corresponding data structures.
+
+    Args:
+        enum_class (Union[Type[Enum], Type[EnumMeta]]): A single Enum class or a Union of Enum classes representing the keys for the generated model.
+        value_model (BaseModel): A Pydantic model representing the structure of the values.
+        model_name (str): The name of the generated Pydantic model.
+
+    Example usage:
+
+    ```python
+    class MyEnum1(str, Enum):
+        key1 = "key1"
+        key2 = "key2"
+
+    class MyEnum2(str, Enum):
+        key3 = "key3"
+        key4 = "key4"
+
+    class ValueModel(BaseModel):
+        value_field: int
+
+    # Create an instance of PydanticModelGenerator with a single Enum
+    model_generator1 = PydanticModelGenerator(MyEnum1, ValueModel, "MyGeneratedModel1")
+
+    # Create an instance of PydanticModelGenerator with a Union of Enums
+    model_generator2 = PydanticModelGenerator(Union[MyEnum1, MyEnum2], ValueModel, "MyGeneratedModel2")
+
+    # Retrieve the generated models using the __call__ method
+    GeneratedModel1 = model_generator1()
+    GeneratedModel2 = model_generator2()
+    ```
+    """
+
+    def __init__(
+        self,
+        enum_class: Union[Type[Enum], Type[EnumMeta]],
+        value_model: BaseModel,
+        model_name: str,
+    ):
+        self.enum_class = enum_class
+        self.value_model = value_model
+        self.model_name = model_name
+        self.generated_model = self._generate_model()
```

Comment:
Instead of storing it, should we just call this method with `__call__`? Or even rename `_generate_model` to  `__call__`.

Reply:
Fixed, thanks 😄